pkg/operator/staticpod/controller/missingstaticpod/missing_static_pod_controller.go

```diff
@@ -0,0 +1,274 @@
+
+// New checks the latest static pods pods and uses the installer to get the
+// latest revision.  If the installer pod was successful, and if it has been
+// longer than the terminationGracePeriodSeconds+10seconds since the installer
+// pod completed successfully, and if the static pod is not at the correct
+// revision, this controller will go degraded.  It will also emit an event for
+// detection in CI.
+func New(
+	operatorClient v1helpers.StaticPodOperatorClient,
+	kubeInformersForTargetNamespace informers.SharedInformerFactory,
+	eventRecorder events.Recorder,
+	targetNamespace string,
+	staticPodName string,
+) factory.Controller {
+	c := &missingStaticPodController{
+		operatorClient:                    operatorClient,
+		podListerForTargetNamespace:       kubeInformersForTargetNamespace.Core().V1().Pods().Lister().Pods(targetNamespace),
+		configMapListerForTargetNamespace: kubeInformersForTargetNamespace.Core().V1().ConfigMaps().Lister().ConfigMaps(targetNamespace),
+		targetNamespace:                   targetNamespace,
+		operandName:                       staticPodName,
+	}
+	return factory.New().
+		ResyncEvery(time.Minute).
+		WithInformers(
+			operatorClient.Informer(),
+			kubeInformersForTargetNamespace.Core().V1().Pods().Informer(),
+		).
+		WithBareInformers(
+			kubeInformersForTargetNamespace.Core().V1().ConfigMaps().Informer(),
+		).
+		WithSync(c.sync).
+		// TODO(dgrisonnet): set degrade status on error once we have more
+		// confidence into this controller.
+		// WithSyncDegradedOnError(operatorClient).
+		ToController("MissingStaticPodController", eventRecorder)
+}
+
+func (c *missingStaticPodController) sync(ctx context.Context, syncCtx factory.SyncContext) error {
+	_, originalOperatorStatus, _, err := c.operatorClient.GetStaticPodOperatorState()
+	if err != nil {
+		return err
+	}
+
+	installerPods, err := c.podListerForTargetNamespace.List(labels.SelectorFromSet(labels.Set{"app": "installer"}))
+	if err != nil {
+		return err
+	}
+
+	// get the most recent installer pod for each node
+	latestInstallerPodsByNode, err := getMostRecentInstallerPodByNode(installerPods)
+	if err != nil {
+		return err
+	}
+
+	errors := make([]string, 0, len(latestInstallerPodsByNode))
+	for node, latestInstallerPodOnNode := range latestInstallerPodsByNode {
+		installerPodRevision, err := getInstallerPodRevision(latestInstallerPodOnNode)
+		if err != nil {
+			// we expect every installer pod to have a installerPodRevision in its name, unexpected error here
+			return fmt.Errorf("failed to get installerPodRevision for installer pod %q - %w", latestInstallerPodOnNode.Name, err)
+		}
+
+		finishedAt, ok := installerPodFinishedAt(latestInstallerPodOnNode)
+		if !ok {
+			// either it's in the process of running, or it ran into an error
+			continue
+		}
+
+		gracePeriod, err := c.getStaticPodTerminationGracePeriodSecondsForRevision(installerPodRevision)
+		if err != nil {
+			return err
+		}
+		threshold := gracePeriod + 10*time.Second
```

Comment:
it looks like `10 s` might be too tight. I have just looked over [this](https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_cluster-kube-controller-manager-operator/605/pull-ci-openshift-cluster-kube-controller-manager-operator-master-e2e-aws-ccm/1493177597461270528) run and it looks like a kcm pod was reported after `102`s, here is a timeline

at `11:40:39` the installer-9-ip-10-0-151-209.ec2.internal finished (Exit code 0)

at `11:40:52` the node agent started deleting the previous instance `"Trying to delete pod" pod="openshift-kube-controller-manager/kube-controller-manager-ip-10-0-151-209.ec2.internal" podUID=1007ded5-acb3-4f6b-8a62-6d08f7a910b9`

at `11:41:49` the node agent reported the new instance as running `"Pod status updated" pod="openshift-kube-controller-manager/kube-controller-manager-ip-10-0-151-209.ec2.internal" status=Running`

at `11:42:01` the installer controller reported `CurrentNode=9` `Updated node "ip-10-0-151-209.ec2.internal" from revision 7 to 9 because static pod is ready`

Reply:
similarly [in ](https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift_cluster-kube-controller-manager-operator/605/pull-ci-openshift-cluster-kube-controller-manager-operator-master-e2e-aws-operator/1493177597528379392/artifacts/e2e-aws-operator/gather-must-gather/artifacts/), a kcm pod was reported after `105 s`:

the installer-pod for revision `6` finished at `2022-02-14T11:30:18Z` and 
the kcm pod was reported at `11:32:03` `Updated node "ip-10-0-231-223.us-east-2.compute.internal" from revision 5 to 6 because static pod is ready`

Reply:
@p0lyn0mial it looks like kcmo does not set `TerminationGracePeriodSeconds` so for those cases we should default to 30s, just like specified in https://github.com/kubernetes/kubernetes/blob/b591acca57f918404f35b8905021f2bda0e42866/staging/src/k8s.io/api/core/v1/types.go#L2722

Reply:
I see that's being handled there, properly, nvmd.